src/platform/library/schedule/ll_schedule.c

```diff
@@ -4,14 +4,302 @@
+		 * to actual DSP FW.
+		 */
+		if (tick_period_us) {
+			while (1) {
+				/* wait for next tick */
+				err = nanosleep(&ts, &ts);
+				if (err == 0)
+					break; /* sleep fully completed */
+				else if (err == EINTR) {
+					continue; /* interrupted - keep going */
+				} else {
+					/* something bad happened ... */
+					fprintf(stderr, "error: sleep failed: %s\n",
+						strerror(err));
+					goto out;
+				}
+			}
+		}
+
+		/* LL time slice now running at this point */
+		pthread_mutex_lock(&vc->list_mutex);
+
+		/* list empty then return */
+		if (list_is_empty(&vc->list)) {
+			pthread_mutex_unlock(&vc->list_mutex);
+			fprintf(stdout, "LL scheduler thread exit - list empty\n");
+			break;
+		}
+
+		/* iterate through the task list */
+		list_for_item_safe(tlist, tlist_, &vc->list) {
+			task = container_of(tlist, struct task, list);
+
+			/* only run queued tasks */
+			if (task->state == SOF_TASK_STATE_QUEUED) {
+				task->state = SOF_TASK_STATE_RUNNING;
+				pthread_mutex_unlock(&vc->list_mutex);
+
+				/* run task and time it */
+				clock_gettime(CLOCK_MONOTONIC, &td0);
+				task->ops.run(task->data);
+				clock_gettime(CLOCK_MONOTONIC, &td1);
+
+				pthread_mutex_lock(&vc->list_mutex);
+
+				/* only re-queue if not cancelled */
+				if (task->state == SOF_TASK_STATE_RUNNING)
+					task->state = SOF_TASK_STATE_QUEUED;
+
+				/* Calculate average task exec time */
+				delta = (td1.tv_sec - td0.tv_sec) * 1000000;
+				delta += (td1.tv_nsec - td0.tv_nsec) / 1000;
+				task->start += delta;
+			}
+		}
+
+		pthread_mutex_unlock(&vc->list_mutex);
+	}
+
+out:
+	/* nothing in list so stop LL thread */
+	vc->vcore_ready = 0;
+	return NULL;
+}
+
+static int schedule_ll_task_complete(void *data, struct task *task)
+{
+	struct ll_vcore *vc = data;
+
+	/* task is complete so remove it from list */
+	pthread_mutex_lock(&vc->list_mutex);
+	list_item_del(&task->list);
+	task->state = SOF_TASK_STATE_COMPLETED;
+	pthread_mutex_unlock(&vc->list_mutex);
+
+	return 0;
+}
+
+/* schedule new LL task */
+static int schedule_ll_task(void *data, struct task *task, uint64_t start,
+			    uint64_t period)
+{
+	struct ll_vcore *vc = data;
+	pthread_attr_t attr;
+	struct sched_param param;
+	int err;
+	bool valid_attr = false;
+	uid_t uid = getuid();
+	uid_t euid = geteuid();
+
+	/* add task to list */
+	pthread_mutex_lock(&vc->list_mutex);
+	list_item_prepend(&task->list, &vc->list);
+	task->state = SOF_TASK_STATE_QUEUED;
+	task->start = 0;
+	pthread_mutex_unlock(&vc->list_mutex);
+
+	/* is vcore thread running ? */
+	if (!vc->vcore_ready) {
+		/* do we have elevated privileges to attempt RT priority */
+		if (uid < 0 || uid != euid) {
+			/* attempt to set thread priority - needs suid */
+			printf("ll schedule: set RT priority\n");
+			err = pthread_attr_init(&attr);
+			if (err) {
+				printf("error: can't create thread attr %d %s\n",
+				       err, strerror(err));
+				goto create;
+			}
+
+			err = pthread_attr_setschedpolicy(&attr, SCHED_FIFO);
```

Comment:
@singalsu @lgirdwood SCHED_FIFO can be quite dangerous in a test bench. We want this for performance tests, but not sure if we want this for general use. OTOH, you can simply not run as root to prevent this now.

Reply:
should I put this under a default-no kconfig?